python/private/bzlmod/python.bzl

```diff
@@ -104,22 +104,28 @@ def _python_impl(module_ctx):
                 is_root = mod.is_root,
             )
 
-            # Only the root module and rules_python are allowed to specify the default
-            # toolchain for a couple reasons:
-            # * It prevents submodules from specifying different defaults and only
-            #   one of them winning.
-            # * rules_python needs to set a soft default in case the root module doesn't,
-            #   e.g. if the root module doesn't use Python itself.
-            # * The root module is allowed to override the rules_python default.
             if mod.is_root:
+                # Only the root module and rules_python are allowed to specify the default
+                # toolchain for a couple reasons:
+                # * It prevents submodules from specifying different defaults and only
+                #   one of them winning.
+                # * rules_python needs to set a soft default in case the root module doesn't,
+                #   e.g. if the root module doesn't use Python itself.
+                # * The root module is allowed to override the rules_python default.
                 # A single toolchain is treated as the default because it's unambiguous.
                 is_default = toolchain_attr.is_default or len(mod.tags.toolchain) == 1
+                # Only the root module should have a say on whether or not to ignore the root user 
+                # error. This can be achieved by taking the ignore_root_user_error for the root 
+                # module. For all other modules, the ignore_root_user_error attr is set to True. 
```

Comment:
Should it be set to the same value as the root module? This comment suggests that only the root module can toggle the behaviour, but the non-root modules will have the toggle set to True. I am wondering, what would be the benefit for setting the non-root module root warning to True by default.

Reply:
as far as I understand, [this](https://github.com/bazelbuild/rules_python/blob/main/python/private/bzlmod/python.bzl#L132) will be called over and over again for every module. Meaning that the check will be run for every toolchain in every module. 
Since there is a [guarantee](https://bazel.build/rules/lib/builtins/module_ctx#modules) that the root module will be checked first, the only way the check will even reach submodules, is if all toolchains in the root module have  `ignore_root_user_error` set to `True`. Which is why it is safe to assume that setting `ignore_root_user_error` to `True` is rendering the submodules redundant in the decision.

Reply:
The reason why I did not simply set it equal to the root module, is that the root module can have multiple toolchains, meaning that the flag can be set to different values for different toolchains, and one `False` value is enough to hit the root_user error.
I could have done something like:
ignore_root_user_error = ignore_root_user_error_toolchain1 && ignore_root_user_error_toolchain2 && ignore_root_user_error_toolchainN I guess, but the result is the same as above.

Reply:
> I am wondering, what would be the benefit for setting the non-root module root warning to True by default.

To clarify (double negatives at play): you mean set `ignore_root_user_error=True` by default? The current default is False.

Following the code comments and PR comments, setting it to True makes the runtime directory non-deterministic because of pyc file creation. This "ignore_root_user_error" name is pretty misleading. A more accurate name would be something like "allow_runtime_to_create_nondeterministic_stdlib_pyc_files_upon_first_use".

This setting controls whether the runtime is made read-only. If it's made readonly, then the runtime can't create pyc files for the stdlib files, and there isn't a nondeterministic pyc problem. It probably also prevents the issue we see on the windows CI machines where the set of files in the runtime is changed mid-execution and/or intermittently picking up the temp pyc files mid-execution.

So I would say making `ignore_root_user_error=True` the default would be a bad idea.

The name of this setting and the logic it affects is kind of strange. I would think it would just change the fail() calls to a print() calls? I wonder if something got lost in some other fix or refactoring...

Reply:
I was not trying to make the` ignore_root_user_error=True` by default, I was simply setting the `ignore_root_user=True` in the submodules only ... Assuming there is at least one toolchain registered in the root module,  setting ` ignore_root_user_error=True`  in all toolchains in the submodules is equivalent to only allowing the root module to set the` ignore_root_user_error`, since one tool chain with ` ignore_root_user_error=False` in the root module is enough to trigger the root user error. 

What I did not think of tho, is the comment you have below 😆 it is of course true that the root module may not register any toolchains, and then this logic will default to True (which is not correct). I will update the code to handle this case and come back 😄 

Thanks for the feedback ! 🙏